rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,238 @@
+  ConfigMap config = 3
+}
+
+```
+
+### Accessing Channels
+
+All the channels needed by a node will be passed as part of the initialisation
+message. The channels are represented as repeating of `SenderChannelInfo`s and
+`ReceiverChannelInfo`s. The keys for each will be the name of the channel. This
+means that channels would need unique names in the manifest file. The downside
+of this approach is that the code needs to be aware of the names of the channels
+in the manifest file.
+
+This is not ideal, but we could work around it by adding a mapping as a config
+entry between internal names that the code understand and the channels names in
+the manifest file. We can then use this to support modification of the manifest
+file without having to recompile the Wasm nodes.
+
+## Planning
+
+The initial implementation will be done as a `planner` example. It will create
+multiple test nodes and channels and connect these together. The example will
+also expose a gRPC method to the client so the client can request the manifest
+file. The client will compare the manifest file to the introspection data to
+test that the nodes were created as expected.
+
+Once the intial implementation is done, some of the common reusable code will be
+moved to the SDK. The fist example of such reusable code is support for node to
+accept an initial proto that contains a list of `Sender`s and `Receiver`s that
+wrap the handles for channels that should be available to the node, and
+potentially additional configuration files.
+
+The next step would be to update the existing examples to reuse the
+configuration-based `planner` node.
+
+## Future Work
+
+### Static analysis tools
+
+We can support application developers in finding potential issues in their
+applications through static analysis tools that can help detect common problems:
+
+- Check that all node and channel names are unique
+- Check that there is only one gRPC invocation receiver for each label
+- Check that there is only one HTTP invocation receiver for each label
+- Check that all nodes can write to handles they receive
+- Check that all nodes can read from handles they receive
+- Hihglight points where data is declassified to `Public`
+- Find nodes that can never receive data (unneeded nodes)
+- Find nodes where data will get stuck (dead ends)
+- Find risky patterns (e.g. diamond shaped graph)
+
+### Dynamic sub-graph creation
+
+The first example will be dynamic creation of user-specific nodes and connect
+them to the appropriate channels in the graph. In future this will be extended
+to support template-based definition of subgraphs. The entire sub-graph will
+then be dynamically generated (e.g. a user-specific sub-graph consisting of
+multiple nodes each).
+
+### Declaration of supported message types
+
+Each node will be able to declare which message types it can read and write.
+Each channel will be annotated with the message types is can handle. This will
```

Comment:
I think we should assume that each channel handles exactly one type, and that the type is expressed as protobuf. If multiple types are needed, they may be wrapped in another type with a oneof, i.e. leveraging sum / union types. Also mention that protobuf types may contain handles to further channels, given our custom support for handles in protobuf that @wildarch implemented.

Reply:
I will add a note that each channel will support a single type. Each node could have multiple input and output channels, so it would still be useful for a node to specify which types it can send and receive to ensure that the channels that are connected can be used by the node.

Reply:
I updated the text to capture the comment about a single message type pe channel.